modules/apps/data-engine/data-engine-taglib/src/main/resources/META-INF/resources/data_layout_builder/js/drag-and-drop/DataLayoutBuilderDragAndDrop.es.js

```diff
@@ -12,83 +12,95 @@
-		'.col-empty .ddm-target:not([data-drop-disabled="true"])'
-	),
-];
-
-const getFields = () => [
-	...document.querySelectorAll(
-		'.ddm-field-container.ddm-target:not([data-drop-disabled="true"])'
-	),
-];
-
-const getColumnKey = (node) => {
-	const {columnIndex, pageIndex, rowIndex} = getIndexes(node.parentElement);
-	const placeholder = !!dom.closest(node, '.placeholder');
-	const parentField = dom.closest(node, '.ddm-field');
-
-	let parentFieldName = '';
-
-	if (parentField) {
-		parentFieldName = getFieldKey(parentField);
-	}
-
-	return `${parentFieldName}_column_${pageIndex}_${rowIndex}_${columnIndex}_${placeholder}`;
-};
-
-const getFieldKey = (node) => {
-	return node.dataset.fieldName;
-};
+import {
+	DRAG_DATA_DEFINITION_FIELD,
+	DRAG_FIELDSET,
+	DRAG_FIELD_TYPE,
+} from './dragTypes.es';
 
 export default ({dataLayoutBuilder}) => {
-	const [columns, setColumns] = useState(getColumns());
-	const [fields, setFields] = useState(getFields());
+	const [{dataDefinition}] = useContext(AppContext);
+
+	const onDrop = useCallback(
+		({data, type}, monitor, {fieldName, origin, ...indexes}) => {
+			if (monitor.didDrop()) {
+				return;
+			}
+
+			let parentFieldName;
+			const parentFieldNode = dom.closest(
+				document.querySelector(`div[data-field-name="${fieldName}"]`),
+				'.ddm-field'
+			);
+
+			if (parentFieldNode) {
+				parentFieldName = parentFieldNode.dataset.fieldName;
+			}
+
+			if (type === DRAG_FIELD_TYPE) {
+				const payload = dropLayoutBuilderField({
+					dataLayoutBuilder,
+					fieldName,
+					fieldTypeName: data.name,
+					indexes,
+					parentFieldName,
+				});
+
+				if (origin === 'empty') {
+					dataLayoutBuilder.dispatch('fieldAdded', payload);
+				}
+				else {
+					dataLayoutBuilder.dispatch('sectionAdded', payload);
+				}
+			}
+			else if (type === DRAG_DATA_DEFINITION_FIELD) {
+				const payload = dropCustomObjectField({
+					dataDefinition,
+					dataDefinitionFieldName: data.name,
+					dataLayoutBuilder,
+					fieldName,
+					indexes,
+					parentFieldName,
+				});
+
+				if (origin === 'empty') {
+					dataLayoutBuilder.dispatch('fieldAdded', payload);
+				}
+				else {
+					dataLayoutBuilder.dispatch('sectionAdded', payload);
+				}
+			}
+			else if (type === DRAG_FIELDSET) {
+				dataLayoutBuilder.dispatch(
+					'fieldSetAdded',
+					dropFieldSet({
+						dataLayoutBuilder,
+						fieldName,
+						fieldSet: data.fieldSet,
+						indexes,
+						parentFieldName,
+					})
+				);
+			}
+		},
+		[dataDefinition, dataLayoutBuilder]
+	);
 
 	useEffect(() => {
-		const provider = dataLayoutBuilder.getLayoutProvider();
-
-		const eventHandler = provider.on('rendered', () => {
-			setColumns(getColumns());
-			setFields(getFields());
-		});
+		const {formBuilderWithLayoutProvider} = dataLayoutBuilder;
 
-		return () => eventHandler.removeListener();
-	}, [dataLayoutBuilder]);
+		formBuilderWithLayoutProvider.props.formBuilderProps.dnd.spec.drop = onDrop;
+	}, [dataLayoutBuilder, onDrop]);
```

Comment:
@brunobasto I still find this `useEffect` here strange, I am wondering if we could pass `onDrop` directly on `DataLayoutBuilder.es.js` but I had realized that we need `dataDefinition`, do we have this information there?

Reply:
Let me check

Reply:
Can we make `drop` an event and make it part of the stuff that dataLayoutBuilder.getProvider() already emmits? That way all reaction to what the form renderer does goes through an event-driven interface

Reply:
One could use it like so
```js
dataLayoutBuilder.getProvider().on('drop', payload => reactToDrop(payload))
```

Reply:
Like we do here: https://github.com/liferay/liferay-portal/blob/master/modules/apps/data-engine/data-engine-taglib/src/main/resources/META-INF/resources/data_layout_builder/js/data-layout-builder/DataLayoutBuilderContextProvider.es.js#L67

Reply:
Hmm, I think we can do that. I'll try later

Reply:
hey @brunobasto I added some improvements to the drop engine in the data engine, now it’s much cleaner and simpler some things.